cpg_pipes/providers/inputs.py

```diff
@@ -0,0 +1,417 @@
+        only_samples: list[str] | None = None,
+        skip_datasets: list[str] | None = None,
+    ) -> list[dict[str, str]]:
+        """
+        Apply the only_samples and skip_samples filters.
+        """
+        filtered_entries = []
+        for entry in entries:
+            cpgid = self.get_sample_id(entry)
+            extid = self.get_external_id(entry)
+            ds = self.get_dataset_name(entry)
+            if skip_datasets and ds and ds in skip_datasets:
+                logger.info(
+                    f'Skipping sample (the dataset is skipped): {ds}|{cpgid}|{extid}'
+                )
+                continue
+            if only_samples:
+                if cpgid in only_samples or extid in only_samples:
+                    logger.info(f'Picking sample: {ds}|{cpgid}|{extid}')
+                else:
+                    continue
+            if skip_samples:
+                if cpgid in skip_samples or extid in skip_samples:
+                    logger.info(f'Skiping sample: {ds}|{cpgid}|{extid}')
+                    continue
+            filtered_entries.append(entry)
+        return filtered_entries
+
+    def _add_sample(self, dataset: Dataset, entry: dict):
+        """
+        Helper method to create a sample in a dataset.
+        """
+        entry = dict(entry)  # making a copy, so methods can call .pop()
+        dataset.add_sample(
+            id=str(self.get_sample_id(entry)),
+            external_id=str(self.get_external_id(entry)),
+            participant_id=self.get_participant_id(entry),
+            sex=self.get_participant_sex(entry),
+            sequencing_type=self.get_sequencing_type(entry),
+            meta=self.get_sample_meta(entry),
+        )
+
+
+class FieldMap(Enum):
+    """
+    CSV field names.
+    """
+
+    dataset = 'dataset'
+    sample = 'sample'
+    external_id = 'external_id'
+    participant_id = 'participant_id'
+    fqs_r1 = 'fqs_r1'
+    fqs_r2 = 'fqs_r2'
+    cram = 'cram'
+    sex = 'sex'
+    sequencing_type = 'seq_type'
+
+
+class CsvInputProvider(InputProvider):
+    """
+    Input provider that parses data from a CSV file.
+    """
+
+    def __init__(self, fp, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+
+        self.d = list(csv.DictReader(fp))
+        if len(self.d) == 0:
+            raise InputProviderError('Empty metadata TSV')
+        if 'sample' not in self.d[0]:
+            raise InputProviderError(
+                f'Metadata TSV must have a header and a column "sample", '
+                f'got: {self.d[0]}'
+            )
+
+    def get_entries(
+        self,
+        dataset: Dataset | None = None,
+    ) -> list[dict]:
+        """
+        Return list of data entries. Optionally, specific for a dataset.
+        """
+        entries = self.d
+        if dataset:
+            entries = [
+                e
+                for e in entries
+                if (
+                    not self.get_dataset_name(e)
```

Comment:
E.g. a CSV file entry doesn't need to have a "dataset" field. It would default to the global "analysis_dataset" (required `--analysis-dataset` option.)